File: frontend/src/services/preferences.ts
export interface UserPreferences {
  torrent_display_mode: "table" | "card" | "card_b" | "list";
  compact: boolean;
  background_image_blur_intensity: number;
  active_color_palette: number;
  // Color Palette 1
  color_palette_1_primary: string;
  color_palette_1_secondary: string;
  color_palette_1_accent: string;
  color_palette_1_muted: string;
  // Color Palette 2
  color_palette_2_primary: string;
  color_palette_2_secondary: string;
  color_palette_2_accent: string;
  color_palette_2_muted: string;
  // Color Palette 3
  color_palette_3_primary: string;
  color_palette_3_secondary: string;
  color_palette_3_accent: string;
  color_palette_3_muted: string;
}

const PREFERENCES_KEY = "user_preferences";

export const preferencesService = {
  // Save preferences to localStorage
  save(preferences: UserPreferences): void {
    try {
      localStorage.setItem(PREFERENCES_KEY, JSON.stringify(preferences));
    } catch (error) {
      console.error("Failed to save preferences to localStorage:", error);
    }
  },

  // Load preferences from localStorage
  load(): UserPreferences | null {
    try {
      const stored = localStorage.getItem(PREFERENCES_KEY);
      if (stored) {
        return JSON.parse(stored);
      }
    } catch (error) {
      console.error("Failed to load preferences from localStorage:", error);
    }
    return null;
  },

  // Clear preferences from localStorage
  clear(): void {
    try {
      localStorage.removeItem(PREFERENCES_KEY);
    } catch (error) {
      console.error("Failed to clear preferences from localStorage:", error);
    }
  },

  // Get default preferences
  getDefaults(): UserPreferences {
    return {
      torrent_display_mode: "card",
      compact: false,
      background_image_blur_intensity: 50,
      active_color_palette: 1,
      color_palette_1_primary: "#3b82f6",
      color_palette_1_secondary: "#8b5cf6",
      color_palette_1_accent: "#10b981",
      color_palette_1_muted: "#6b7280",
      color_palette_2_primary: "#ef4444",
      color_palette_2_secondary: "#f59e0b",
      color_palette_2_accent: "#ec4899",
      color_palette_2_muted: "#78716c",
      color_palette_3_primary: "#06b6d4",
      color_palette_3_secondary: "#14b8a6",
      color_palette_3_accent: "#a855f7",
      color_palette_3_muted: "#64748b",
    };
  },
};

File: backend/internal/mappers/preferences.go
package mappers

import (
	"github.com/google/uuid"
	"github.com/jfxdev/gardarr/internal/models"
	"github.com/jfxdev/gardarr/internal/schemas"
)

// ToPreferencesResponse converts UserPreferences model to PreferencesResponse
func ToPreferencesResponse(prefs models.UserPreferences) models.PreferencesResponse {
	return models.PreferencesResponse{
		TorrentDisplayMode:           prefs.TorrentDisplayMode,
		Compact:                      prefs.Compact,
		BackgroundImageBlurIntensity: prefs.BackgroundImageBlurIntensity,
		ActiveColorPalette:           prefs.ActiveColorPalette,
		ColorPalette1Primary:         prefs.ColorPalette1Primary,
		ColorPalette1Secondary:       prefs.ColorPalette1Secondary,
		ColorPalette1Accent:          prefs.ColorPalette1Accent,
		ColorPalette1Muted:           prefs.ColorPalette1Muted,
		ColorPalette2Primary:         prefs.ColorPalette2Primary,
		ColorPalette2Secondary:       prefs.ColorPalette2Secondary,
		ColorPalette2Accent:          prefs.ColorPalette2Accent,
		ColorPalette2Muted:           prefs.ColorPalette2Muted,
		ColorPalette3Primary:         prefs.ColorPalette3Primary,
		ColorPalette3Secondary:       prefs.ColorPalette3Secondary,
		ColorPalette3Accent:          prefs.ColorPalette3Accent,
		ColorPalette3Muted:           prefs.ColorPalette3Muted,
	}
}

// GetDefaultPreferences returns a new UserPreferences model with default values
func GetDefaultPreferences(userUUID uuid.UUID) models.UserPreferences {
	return models.UserPreferences{
		UUID:                         uuid.New(),
		UserUUID:                     userUUID,
		TorrentDisplayMode:           "card",
		Compact:                      false,
		BackgroundImageBlurIntensity: 50,
		ActiveColorPalette:           1,
		ColorPalette1Primary:         "#3b82f6",
		ColorPalette1Secondary:       "#8b5cf6",
		ColorPalette1Accent:          "#10b981",
		ColorPalette1Muted:           "#6b7280",
		ColorPalette2Primary:         "#ef4444",
		ColorPalette2Secondary:       "#f59e0b",
		ColorPalette2Accent:          "#ec4899",
		ColorPalette2Muted:           "#78716c",
		ColorPalette3Primary:         "#06b6d4",
		ColorPalette3Secondary:       "#14b8a6",
		ColorPalette3Accent:          "#a855f7",
		ColorPalette3Muted:           "#64748b",
	}
}

// GetDefaultPreferencesResponse returns a PreferencesResponse with default values
func GetDefaultPreferencesResponse() models.PreferencesResponse {
	defaults := GetDefaultPreferences(uuid.Nil)
	return ToPreferencesResponse(defaults)
}

// ApplyPreferenceUpdates applies updates from request to preferences model
func ApplyPreferenceUpdates(prefs *models.UserPreferences, req schemas.UpdatePreferencesRequest) {
	if req.TorrentDisplayMode != nil {
		prefs.TorrentDisplayMode = *req.TorrentDisplayMode
	}
	if req.Compact != nil {
		prefs.Compact = *req.Compact
	}
	if req.BackgroundImageBlurIntensity != nil {
		prefs.BackgroundImageBlurIntensity = *req.BackgroundImageBlurIntensity
	}
	if req.ActiveColorPalette != nil {
		prefs.ActiveColorPalette = *req.ActiveColorPalette
	}
	// Color Palette 1
	if req.ColorPalette1Primary != nil {
		prefs.ColorPalette1Primary = *req.ColorPalette1Primary
	}
	if req.ColorPalette1Secondary != nil {
		prefs.ColorPalette1Secondary = *req.ColorPalette1Secondary
	}
	if req.ColorPalette1Accent != nil {
		prefs.ColorPalette1Accent = *req.ColorPalette1Accent
	}
	if req.ColorPalette1Muted != nil {
		prefs.ColorPalette1Muted = *req.ColorPalette1Muted
	}
	// Color Palette 2
	if req.ColorPalette2Primary != nil {
		prefs.ColorPalette2Primary = *req.ColorPalette2Primary
	}
	if req.ColorPalette2Secondary != nil {
		prefs.ColorPalette2Secondary = *req.ColorPalette2Secondary
	}
	if req.ColorPalette2Accent != nil {
		prefs.ColorPalette2Accent = *req.ColorPalette2Accent
	}
	if req.ColorPalette2Muted != nil {
		prefs.ColorPalette2Muted = *req.ColorPalette2Muted
	}
	// Color Palette 3
	if req.ColorPalette3Primary != nil {
		prefs.ColorPalette3Primary = *req.ColorPalette3Primary
	}
	if req.ColorPalette3Secondary != nil {
		prefs.ColorPalette3Secondary = *req.ColorPalette3Secondary
	}
	if req.ColorPalette3Accent != nil {
		prefs.ColorPalette3Accent = *req.ColorPalette3Accent
	}
	if req.ColorPalette3Muted != nil {
		prefs.ColorPalette3Muted = *req.ColorPalette3Muted
	}
}
